File: leetcode/C++/0003-m-longest-substring-without-repeating-characters.cpp
// Two pointers
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        vector<int> letters(256, 0);
        int res = 0;
        int i = 0, j = 0;
        while (i <= j && j < n) {
            while (letters[(int)s[j]] == 0 && j < n) {
                letters[(int)s[j]]++;
                j++;
            }
            res = max(res, j - i);
            letters[(int)s[i]]--;
            i++;
        }
        return res;
    }
};

File: leetcode/Python/0003-m-longest-substring-without-repeating-characters.py
"""
Time: O(n)
Space: O(1)
"""
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        res = 0
        map = {}
        for i in range(len(s)):
            if s[i] in map and map[s[i]] >= left:
                left = map[s[i]] + 1
            map[s[i]] = i
            res = max(res, i - left + 1)
        return res